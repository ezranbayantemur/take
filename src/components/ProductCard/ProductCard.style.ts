import {StyleSheet, Dimensions} from 'react-native';
import {color, font, padding, typography} from '@style';

const deviceSize = Dimensions.get('window');

const CONTAINER_WIDTH_LIMIT = 0.45;
const IMAGE_WIDTH_LIMIT = 0.4;

export default StyleSheet.create({
  container: {
    marginVertical: padding.MEDIUM,
    borderWidth: 1,
    borderColor: color.MERCURY,
    borderRadius: 10,
    maxWidth: deviceSize.width * CONTAINER_WIDTH_LIMIT,
    alignItems: 'center',
  },
  image: {
    maxWidth: deviceSize.width * IMAGE_WIDTH_LIMIT,
    width: 155,
    height: 185,
  },
  info_container: {
    alignItems: 'center',
    padding: padding.SMALL,
  },
  name: {
    fontFamily: font.REGULAR,
    fontSize: typography.NORMAL,
  },
  seller: {
    fontSize: typography.NORMAL,
    fontFamily: font.BOLD,
    color: color.SMOKE,
  },
  price: {
    marginTop: padding.TINY,
    fontSize: typography.NORMAL,
    fontFamily: font.BOLD,
  },
});
