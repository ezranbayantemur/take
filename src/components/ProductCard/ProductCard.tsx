import {View, Text, Image, TouchableOpacity} from 'react-native';
import React from 'react';
import type {ProductCardProps} from './ProductCard.types';

import styles from './ProductCard.style';

const ProductCard = ({testID, data, onSelect}: ProductCardProps) => {
  const {product_image, product_name, product_price, seller_name} = data;

  const handleOnSelect = () => {
    onSelect(data);
  };

  return (
    <TouchableOpacity
      testID={`${testID}_productcard_touchable`}
      style={styles.container}
      onPress={handleOnSelect}>
      <Image
        style={styles.image}
        source={{uri: product_image}}
        resizeMode="contain"
      />
      <View style={styles.info_container}>
        <Text style={styles.name}>{product_name}</Text>
        <Text style={styles.seller}>{seller_name}</Text>
        <Text style={styles.price}>{product_price} TL</Text>
      </View>
    </TouchableOpacity>
  );
};

export default ProductCard;
