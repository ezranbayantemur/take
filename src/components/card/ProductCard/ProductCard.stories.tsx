import {ProductCard} from '@components';
import productsData from '@mocks/products_for_category.json';
import type {Meta, StoryObj} from '@storybook/react-native';
import type {Product} from '@types';
import type {ProductCardProps} from './ProductCard.types';

const mockData: Product = productsData[0];

const ProductCardMeta: Meta<ProductCardProps> = {
  args: {
    data: mockData,
  },
  title: 'ProductCard',
  component: ProductCard,
};

export default ProductCardMeta;

export const Basic: StoryObj<ProductCardProps> = {};
