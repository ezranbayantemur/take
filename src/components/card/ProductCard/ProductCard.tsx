import {View, Text, Image, TouchableOpacity} from 'react-native';
import React from 'react';
import styles from './ProductCard.style';
import type {ProductCardProps} from './ProductCard.types';

const ProductCard = ({testID, data, onSelect}: ProductCardProps) => {
  const {product_image, product_name, product_price, seller_name} = data;

  const handleOnSelect = () => {
    onSelect(data);
  };

  return (
    <TouchableOpacity
      testID={`${testID}_productcard_touchable`}
      style={styles.container}
      onPress={handleOnSelect}>
      <View style={styles.image_container}>
        <Image
          style={styles.image}
          source={{uri: product_image}}
          resizeMode="contain"
        />
      </View>
      <View style={styles.info_container}>
        <Text style={styles.name}>{product_name}</Text>
        <Text style={styles.seller}>{seller_name}</Text>
        <Text style={styles.price}>
          {product_price.toLocaleString('tr-TR')} TL
        </Text>
      </View>
    </TouchableOpacity>
  );
};

export default React.memo(ProductCard);
