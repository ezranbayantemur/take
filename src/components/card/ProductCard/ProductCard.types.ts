import type {Product} from '@types';

export interface ProductCardProps {
  testID: string;
  data: Product;
  onSelect: (product: Product) => void;
}
