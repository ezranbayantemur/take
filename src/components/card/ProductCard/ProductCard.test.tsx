import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import productData from '@mocks/products_for_category.json';
import ProductCard from './ProductCard';
import type {Product} from '@types';

const mockData: Product = productData[0];
const mockOnSelect = jest.fn();

let wrapper: ReturnType<typeof render>;

describe('ProductCard unit tests', () => {
  beforeEach(() => {
    wrapper = render(
      <ProductCard testID="test" data={mockData} onSelect={mockOnSelect} />,
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render texts correctly', () => {
    expect(wrapper.queryByText(mockData.product_name)).not.toBeNull();
    expect(wrapper.queryByText(mockData.seller_name)).not.toBeNull();
    expect(wrapper.queryByText(`${mockData.product_price} TL`)).not.toBeNull();
  });

  it('should trigger onSelect correctly', () => {
    const touchable = wrapper.getByTestId('test_productcard_touchable');
    fireEvent(touchable, 'onPress');

    expect(mockOnSelect).toBeCalled();
  });
});
