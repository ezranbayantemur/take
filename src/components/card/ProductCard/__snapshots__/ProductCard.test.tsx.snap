// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard unit tests should render correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FBFBFB",
      "borderColor": "#E1E1E1",
      "borderRadius": 10,
      "borderWidth": 1,
      "flex": 1,
      "marginVertical": 16,
      "maxWidth": 337.5,
      "opacity": 1,
    }
  }
  testID="test_productcard_touchable"
>
  <View
    style={
      {
        "padding": 8,
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        {
          "uri": "https://cdn.tmobile.com/content/dam/t-mobile/en-p/cell-phones/apple/Apple-iPhone-14-Pro/Deep-Purple/Apple-iPhone-14-Pro-Deep-Purple-thumbnail.png",
        }
      }
      style={
        {
          "height": 185,
          "maxWidth": 300,
          "padding": 8,
          "width": 155,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "padding": 8,
      }
    }
  >
    <Text
      style={
        {
          "fontFamily": "Lato-Regular",
          "fontSize": 15,
        }
      }
    >
      iPhone 14 Pro
    </Text>
    <Text
      style={
        {
          "color": "#BDBDBD",
          "fontFamily": "Lato-Bold",
          "fontSize": 15,
        }
      }
    >
      Apple
    </Text>
    <Text
      style={
        {
          "fontFamily": "Lato-Bold",
          "fontSize": 15,
          "marginTop": 4,
        }
      }
    >
      400
       TL
    </Text>
  </View>
</View>
`;
